core/main/src/broker/provider_broker_state.rs

```diff
@@ -0,0 +1,126 @@
+// Copyright 2023 Comcast Cable Communications Management, LLC
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+// http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+//
+// SPDX-License-Identifier: Apache-2.0
+//
+
+use std::{
+    collections::HashMap,
+    sync::{Arc, RwLock},
+};
+
+use ripple_sdk::{
+    api::{
+        firebolt::fb_capabilities::FireboltPermission,
+        gateway::rpc_gateway_api::{ApiMessage, ApiProtocol, RpcRequest},
+    },
+    log::{debug, error},
+    tokio
+};
+
+use serde_json::json;
+
+use crate::state::session_state::Session;
+
+use super::endpoint_broker::BrokerRequest;
+
+#[derive(Debug, Clone)]
+pub struct ProvideBrokerState {
+    capability_map: Arc<RwLock<HashMap<String, Session>>>,
+}
+
+pub enum ProviderResult {
+    Session(Session),
+    Registered,
+    NotAvailable(String),
+}
+
+impl ProvideBrokerState {
+    pub fn new() -> Self {
+        Self {
+            capability_map: Arc::new(RwLock::new(HashMap::new())),
+        }
+    }
+    pub fn check_provider_request(
+        &self,
+        request: &RpcRequest,
+        permission: &Vec<FireboltPermission>,
+        session: Option<Session>,
+    ) -> Option<ProviderResult> {
+        debug!("Method {}", request.method);
+        if request.method.contains(".provide") {
```

Comment:
Is there a better way to determine this instead of mandating that a method starts with `.provide`, e.g. can we look up a tag in the schema?

Reply:
Its a combination of .provide and a firebolt capability with a role of Provides
`if let Some(p) = Self::get_permission(permission) {`

Reply:
Maybe in another iteration we can use the RPC validator to query for the tag.